File: Recursion&Backtracking/박연주/14888.py
def dfs(depth, result, add, sub, mul, div):
    global max_value, min_value
    
    # 최대 최소 저장하고 리턴
    if depth == list_size:
        max_value = max(result, max_value)
        min_value = min(result, min_value)
        return
    
    if add:
        dfs(depth+1, result+input_list[depth], add-1, sub, mul, div)
    if sub:
        dfs(depth+1, result-input_list[depth], add, sub-1 , mul, div)
    if mul:
        dfs(depth+1, result*input_list[depth], add, sub, mul-1, div)
    if div:
        dfs(depth+1, int(result/input_list[depth]), add, sub, mul, div-1)
    
        
    
list_size = int(input())
input_list = [int(i) for i in input().split()]
operator_list = [int(i) for i in input().split()]

max_value = -1e9
min_value = 1e9

dfs(1, input_list[0], operator_list[0], operator_list[1], operator_list[2], operator_list[3])

print(max_value)
print(min_value)

File: Recursion&Backtracking/hwichon/14888.kt
import java.io.BufferedReader
import java.io.InputStreamReader

var maxValue = Int.MIN_VALUE
var minValue = Int.MAX_VALUE

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val N = readLine().toInt()
    val numbers = readLine().split(" ").map { it.toInt() }.toIntArray()
    val (add, sub, mul, div) = readLine().split(" ").map { it.toInt() }
    dfs(1, numbers[0], add, sub, mul, div, numbers)
    println(maxValue)
    println(minValue)
}

fun dfs(index: Int, result: Int, add: Int, sub: Int, mul: Int, div: Int, numbers: IntArray) {
    if (index == numbers.size) {
        maxValue = maxOf(maxValue, result)
        minValue = minOf(minValue, result)
        return
    }
    if (add > 0) dfs(index + 1, result + numbers[index], add - 1, sub, mul, div, numbers)
    if (sub > 0) dfs(index + 1, result - numbers[index], add, sub - 1, mul, div, numbers)
    if (mul > 0) dfs(index + 1, result * numbers[index], add, sub, mul - 1, div, numbers)
    if (div > 0) {
        val divResult = if (result < 0) -(-result / numbers[index]) else result / numbers[index]
        dfs(index + 1, divResult, add, sub, mul, div - 1, numbers)
    }
}
